src/manifold/src/manifold.cpp

```diff
@@ -767,4 +830,104 @@ Manifold Manifold::TrimByPlane(glm::vec3 normal, float originOffset) const {
 }
 
 ExecutionParams& ManifoldParams() { return params; }
+
+Manifold Manifold::Hull(const std::vector<float>& vertProps, const int numProp,
+                        const double precision) {
+  const int len = vertProps.size() / numProp;
+  // FIXME: Should there be a hull specific error?
+  if (len < 4) return Invalid();
+  std::vector<glm::vec3> pts(len);
+  for (int i = 0; i < len; i++) {
+    const int j = i * numProp;
+    pts[i] = {vertProps[j], vertProps[j + 1], vertProps[j + 2]};
+  }
+  const auto trip = NonCollinearTriple(pts, precision);
+  const auto plane = PlaneOfPts(pts[trip.x], pts[trip.y], pts[trip.z]);
+  const auto [d, dIsAboveTrip] = NonCoplanar(pts, plane, precision);
+  if (d < 0) return Invalid();  // All points coplanar.
+  const int a = trip.x;
+  const int b = dIsAboveTrip ? trip.z : trip.y;
+  const int c = dIsAboveTrip ? trip.y : trip.z;
+  std::vector<glm::ivec3> triangles;
+  std::vector<glm::vec4> planes;
+  std::unordered_set<glm::ivec2, IVec2Hash> halfEdges{};
+  std::vector<int> dropped;
+  std::vector<bool> kept;
+  const auto AddTri = [&pts, &triangles, &planes, &dropped, &kept](int a, int b,
+                                                                   int c) {
+    auto tri = glm::ivec3(c, b, a);
+    auto plane = PlaneOfPts(pts[a], pts[b], pts[c]);
+    if (dropped.size() > 0) {
+      const int idx = dropped[dropped.size() - 1];
+      triangles[idx] = tri;
+      planes[idx] = plane;
+      kept[idx] = true;
+      dropped.pop_back();
+    } else {
+      triangles.push_back(tri);
+      planes.push_back(plane);
+      kept.push_back(true);
+    }
+  };
+  AddTri(a, b, c);
+  AddTri(d, b, a);
+  AddTri(c, d, a);
+  AddTri(b, d, c);
+  for (int i = 0; i < len; i++) {
+    if (i == a || i == b || i == c || i == d) continue;  // skip starting points
+    // collect half edges of triangles that are in conflict with the points at
+    // idx, pruning the conflicting triangles and their planes in the process
+    for (int j = 0; j < triangles.size(); j++) {
+      if (kept[j] && DistanceToPlane(pts[i], planes[j]) > precision) {
+        halfEdges.insert({triangles[j].x, triangles[j].z});
+        halfEdges.insert({triangles[j].z, triangles[j].y});
+        halfEdges.insert({triangles[j].y, triangles[j].x});
+        // mark triangle/plane as dropped (slot to be reused)
+        dropped.push_back(j);
+        kept[j] = false;
+      }
+    }
+    // form new triangles with the outer perimeter (horizon) of the set of
+    // conflicting triangles and the point at idx
+    for (auto e : halfEdges) {
+      if (halfEdges.erase({e.y, e.x}) == 0) AddTri(e.x, e.y, i);
+    }
```

Comment:
Is this O(n^2)? Or greater? I believe a hull can be O(n log n). How does this compare to others like qhull for performance?

Reply:
The algorithm as a whole may be greater. After I got this one working I started reading more and have some things to try.

Basically, the thing that takes all the time here is checking if each point is outside of the existing triangles in the first loop. There are some optimizations that cut down on the amount of triangles that need to be checked by throwing out points based on a rough low triangle approximations. I'll see if I can find more detail on those modern ones and adapt this or rewrite accordingly.

Reply:
Yeah, but n^2 is a non-starter compared to nlogn, that could easily be a factor of a million for us. I'm pretty sure the efficient algorithms for this are all sweep planes. Before we review this any further, let's compare with something we can take a dependency on (which doesn't include the one you linked, because it has no license), probably https://github.com/qhull/qhull

Reply:
Hmm, those further links look more promising 👍

Reply:
I don't disagree, that's why I've been looking around. I did notice the lack of license when I came back to look. I didn't really think about how bad the one I ported was until I was trying bigger inputs, since my main use of it was doing simple hulls over points to join things together, not trying to hull over multiple shapes (would use OpenSCAD for that since I was generating SCAD files).